src/optical_spi.c

```diff
@@ -2,143 +2,146 @@
-    EICRA |= (1 << ISC10) | (1 << ISC11);
-    // enable external interrupt 0
-    EIMSK |= (1 << INT1);
-    // enable global interrupts
-    sei();
-
     // Initialize PAY-Optical CS pin
     init_cs(OPT_CS_PIN, &OPT_CS_DDR );
     set_cs_high(OPT_CS_PIN, &OPT_CS_PORT);
 
     // Initialize PAY-Optical reset pin
     init_cs(OPT_RST_PIN, &OPT_RST_DDR);
     set_cs_high(OPT_RST_PIN, &OPT_RST_PORT);
+
+    // Initialize PAY-Optical DATA_RDY pin
+    OPT_DATA_DDR &= ~_BV(OPT_DATA);     // write 0, set as input pin
+    OPT_DATA_PORT |= _BV(OPT_DATA);     // activate input pullup resistor
+
 }
 
-/*
-Resets the PAY-Optical microcontroller.
-*/
+
+// Resets the PAY-Optical microcontroller
 void rst_opt_spi(void) {
     // TODO - check how many cycles is necessary
     set_cs_low(OPT_RST_PIN, &OPT_RST_PORT);
     _delay_ms(1000);
     set_cs_high(OPT_RST_PIN, &OPT_RST_PORT);
 }
 
-// Reads 24 bits of raw data from the PAY-Optical microcontroller over optical SPI
-// (Sends the request command, then blocks until it gets 3 bytes back)
-uint32_t read_opt_spi(uint8_t field_num) {
-    // Must make this block non-atomic to ensure interrupts are enabled, in case
-    // this function executes within an atomic block (e.g. in main() of main_test)
-    NONATOMIC_BLOCK(NONATOMIC_RESTORESTATE) {
-        send_opt_spi_cmd(field_num);
-
-        // Wait until we get 3 bytes, with a timeout
-        // In testing, timeout always ended at the original value minus 7 or 8
-        uint32_t timeout = UINT32_MAX;
-        while ((timeout > 0) && (opt_spi_num_bytes < 3)) {
-            timeout--;
-        }
-        print("timeout = %u\n", timeout);
-
-        // If we didn't get 3 bytes
-        if (opt_spi_num_bytes < 3) {
-            //_delay_ms(5000); // long delay = timeout, used for testing
-            return 0;
-        }
 
-        // If we have received all 3 bytes now
-        uint32_t read_data = opt_spi_data;
-        // print ("Received optical data: 0x%06lx\n", read_data);
+// returns status of DATA_RDY pin
+uint8_t get_data_pin(void){
+    return OPT_DATA_PIN >> OPT_DATA;
+}
 
-        // Clear SPI request variables
-        opt_spi_num_bytes = 0;
-        opt_spi_data = 0;
 
-        return read_data;
-    }
+// sends 2 bytes to request reading
+// byte 1 = cmd_byte
+// byte 2 = well_into
+void send_opt_spi_cmd(uint8_t cmd_opcode, uint8_t well_info) {
+    // print("Sending optical the command code %X\n", cmd_opcode);
 
-    // Silence missing return warning
-    return 0;
-}
+    opt_spi_data = 0;
+    spi_frame_number = 0;
 
-/*
-Sends a command to read data for the specified field number from PAY-Optical.
-*/
-void send_opt_spi_cmd(uint8_t field_num) {
-    // print("Sending optical command - field #%u\n", field_num);
+    // Send the command to PAY-Optical to start reading data    
+    set_cs_low(OPT_CS_PIN, &OPT_CS_PORT);
+    send_spi(cmd_opcode);
+    set_cs_high(OPT_CS_PIN, &OPT_CS_PORT);
```

Comment:
Add a 1ms delay here to give time just in case optical doesn't have time to set DATA_RDYn high before setting it low.